core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/flow/ClusterOperationService.java

```diff
@@ -316,55 +320,75 @@ private void handleHealthChange(Set<String> failedNodes, Set<String> newHealthyN
 
     private void handleChangedHosts(Cluster cluster, Set<String> newHealthyNodes,
             Map<String, List<String>> autoRecoveryNodesMap, Map<String, InstanceMetaData> autoRecoveryHostMetadata,
-            Map<String, InstanceMetaData> failedHostMetadata) {
+            Map<InstanceMetaData, Optional<String>> failedHostMetadata) {
         try {
             boolean hasAutoRecoverableNodes = !autoRecoveryNodesMap.isEmpty();
             if (hasAutoRecoverableNodes) {
                 flowManager.triggerClusterRepairFlow(cluster.getStack().getId(), autoRecoveryNodesMap, false, false);
-                updateChangedHosts(cluster, autoRecoveryHostMetadata.keySet(), Set.of(SERVICES_HEALTHY), InstanceStatus.WAITING_FOR_REPAIR,
-                        CLUSTER_AUTORECOVERY_REQUESTED);
+                updateChangedHosts(cluster, autoRecoveryHostMetadata.keySet().stream().collect(Collectors.toMap(host -> host, host -> Optional.empty())),
```

Comment:
all the 3 `updateChangedHosts` calls and parameters are hard to follow and read. I would suggest to move all of them to separate methods, move the paramaters into variables in those methods.

also if `updateChangedHosts` handles when the 2nd parameter is empty we won't need those `if` condition like `!failedHostMetadata.isEmpty()`